src/i18n.ts

```diff
@@ -0,0 +1,259 @@
+import {
+  Context,
+  Fluent,
+  type FluentBundleOptions,
+  FluentType,
+  type FluentValue,
+  type LocaleId,
+  type MiddlewareFn,
+  type NextFunction,
+  resolve,
+  type TranslationContext,
+} from "./deps.ts";
+
+import { readLocalesDir } from "./utils.ts";
+
+import type {
+  I18nConfig,
+  I18nContextFlavor,
+  TranslateFunction,
+} from "./types.ts";
+
+class FluentContext extends FluentType<string> {
+  constructor(value: string, private defaultValue = "") {
+    super(value);
+  }
+
+  // deno-lint-ignore no-explicit-any
+  toString(scope: any): string {
+    let ctx = JSON.parse(scope.args?.ctx as string);
+    const keys = this.value.split(".");
+    for (const key of keys) {
+      ctx = ctx[key];
+      if (ctx === undefined) break;
+    }
+    return ctx ?? this.defaultValue;
+  }
+}
+
+/**
+ * A custom Fluent function for accessing `Context` properties directly from
+ * Fluent translation files.
+ */
+export function CTX(args: FluentValue[]) {
+  if (args[1] && typeof args[1] !== "string") {
+    throw new TypeError("CTX: Only string type is allowed for defaultValue.");
+  }
+
+  const key = args[0];
+  const defaultValue = typeof args[1] === "string" ? args[1] : undefined;
+
+  if (typeof key === "string") {
+    return new FluentContext(key, defaultValue);
+  }
+
+  throw new TypeError(
+    "CTX: Invalid argument type. Both key and defaultValue should be the type of string.",
```

Comment:
The other case was caught earlier, no?
```suggestion
    "CTX: Only type string is allowed for key.",
```